core/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -91,6 +96,43 @@ private[spark] abstract class YarnSchedulerBackend(
   }
 
   /**
+   * Override the DriverEndpoint to add extra logic for the case when an executor is disconnected.
+   * We should check the cluster manager and find if the loss of the executor was caused by YARN
+   * force killing it due to preemption.
+   */
+  private class YarnDriverEndpoint(rpcEnv: RpcEnv, sparkProperties: ArrayBuffer[(String, String)])
+      extends DriverEndpoint(rpcEnv, sparkProperties) {
+
+    /**
+     * When onDisconnected is received at the driver endpoint, the superclass DriverEndpoint
+     * handles it by assuming the Executor was lost for a bad reason and removes the executor
+     * immediately.
+     *
+     * In YARN's case however it is crucial to talk to the application master and ask why the
+     * executor had exited. In particular, the executor may have exited due to the executor
+     * having been preempted. If the executor "exited normally" according to the application
+     * master then we pass that information down to the TaskSetManager to inform the
+     * TaskSetManager that tasks on that lost executor should not count towards a job failure.
+     */
+    override def onDisconnected(rpcAddress: RpcAddress): Unit = {
+      addressToExecutorId.get(rpcAddress).foreach { executorId =>
```

Comment:
Man this is pretty hard to read. I wonder if something like this would compile:

      addressToExecutorId.get(rpcAddress).foreach { executorId =>
        yarnSchedulerEndpoint.askForExecutorLossReason(executorId, 
          reason => driverEndpoint.askWithRetry[Boolean](RemoveExecutor(executorId, reason),
          e => {
            logWarning(s"Attempted to get executor loss reason" +
              s" for $rpcAddress but got no response. Marking as slave lost.", e)
            driverEndpoint.askWithRetry[Boolean](RemoveExecutor(executorId, SlaveLost()))
          })
      }

Since this is the only call site for `askForExecutorLossReason` that I can see, it might be better to just inline the code. Should make it more readable.